modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookupUtility.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless.lookup;
+
+import org.objectweb.asm.Type;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Map;
+
+public final class PainlessLookupUtility {
+
+    // The following terminology is used for variable names throughout the lookup package:
+    //
+    // - javaClass           (Class)         - a java class including def and excluding array type java classes
+    // - javaClassName       (String)        - the fully qualified java class name for a javaClass
+    // - painlessClassName   (String)        - the fully qualified painless name or imported painless name for a painlessClass
+    // - anyClassName        (String)        - either a javaClassName or a painlessClassName
+    // - javaType            (Class)         - a java class excluding def and array type java classes
+    // - painlessType        (Class)         - a java class including def and array type java classes
+    // - javaTypeName        (String)        - the fully qualified java Type name for a javaType
+    // - painlessTypeName    (String)        - the fully qualified painless name or imported painless name for a painlessType
+    // - anyTypeName         (String)        - either a javaTypeName or a painlessTypeName
+    // - painlessClass       (PainlessClass) - a painless class object
+    //
+    // Under ambiguous circumstances most variable names are prefixed with asm, java, or painless.
+    // If the variable name is the same for asm, java, and painless, no prefix is used.
+
+    public static Class<?> javaObjectTypeToPainlessDefType(Class<?> javaType) {
```

Comment:
a brief one liner would be nice on most of these methods, especially since at a glance, you would this this would always return `def.class`, but an explanation of the array case necessary would be nice. These could be in a followup since I know they are just moving here and were undocumented in their previous location.

Reply:
Yes, once I'm done with the functional changes I'll go back and ensure everything is documented.